Containerd Project Security Self-Assessment - Security Pals (#1202)
cncf/tag-security: Containerd/self-assessment.md

Reviewer: Please go back to ensure that no template text like this is included here or in the rest of the document

# Self-assessment

> The Self-assessment is the initial document for projects to begin thinking about the
> security of the project, determining gaps in their security, and preparing any security
> documentation for their users. This document is ideal for projects currently in the
> CNCF **sandbox** as well as projects that are looking to receive a joint assessment and
> currently in CNCF **incubation**.
> 
> # Self-assessment outline

Reviewer 2: This is superflous, see [this example](https://github.com/cncf/tag-security/pull/1181/files) for what to remove

```suggestion
```

## Table of contents

* [Metadata](#metadata)
  * [Security links](#security-links)
* [Overview](#overview)
  * [Actors](#actors)
  * [Actions](#actions)
  * [Background](#background)
  * [Goals](#goals)
  * [Non-goals](#non-goals)
* [Self-assessment use](#self-assessment-use)
* [Security functions and features](#security-functions-and-features)
* [Project compliance](#project-compliance)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
* [Appendix](#appendix)

## Metadata

> A table at the top for quick reference information, later used for indexing.

Reviewer 2: ```suggestion
```

|   |  |
| -- | -- |
| Software | [containerd](https://github.com/containerd/containerd)  |
| Security Provider | No  |
| Languages | Go, C++ |
| SBOM | [Packages](https://github.com/containerd/containerd/tree/main/pkg) [Versions](https://github.com/containerd/containerd/tree/main/version) |
| | |

### Security links

> Provide the list of links to existing security documentation for the project. You may
> use the table below as an example:

Reviewer 2: ```suggestion
```

| Doc | url |
| -- | -- |
| Security file | https://github.com/containerd/project/blob/main/SECURITY.md |
| Default and optional configs | https://github.com/containerd/containerd/blob/main/docs/man/containerd-config.toml.5.md  https://github.com/containerd/containerd/blob/main/docs/cri/config.md  https://github.com/containerd/containerd/blob/main/docs/hosts.md |

## Overview

> Containerd is a Cloud Native Computing Foundation (CNCF) Project focused on providing the core functionalities for container orchestration. Specifically architected to focus on modularity and compatibility, this provides a secure and minimal approach making it a great option for integrating into different container systems. 

Reviewer 2: ```suggestion
Containerd is a container runtime focused on providing the core functionalities for managing container lifecycles. Specifically architected to focus on modularity and compatibility, it provides a secure and minimal approach making it a great option for integrating into different container orchestrators. 
```

![Sample Image](https://github.com/containerd/containerd/blob/main/docs/historical/design/architecture.png)

### Background

Containerd, a fundamental tool in the realm of containerization, provides a dependable and standardized approach to managing containers. It is a lightweight yet powerful container runtime, ensuring a consistent and efficient experience.

> Originally developed by Docker, Inc. as an integral part of the Docker project, Containerd has evolved with the dynamic container ecosystem. Docker's decision to separate container runtime functionality led to Containerd, an independent project dedicated to container management.

Reviewer 2: ```suggestion
Originally developed by Docker, Inc. as an integral part of the Docker project, containerd has evolved with the dynamic container ecosystem. Docker's decision to separate container runtime functionality from the runc project led to containerd, an independent project dedicated to container management.
```

#### Core Features:

**- Image and Container Management:**

Containerd oversees the entire lifecycle of containers, handling tasks such as image storage, transfer, execution, and supervision. Its capabilities also extend to other essential operations like pushing, pulling, and managing container images.

**- Pluggable Architecture:**

Containerd boasts a modular and adaptable architecture, allowing for the assembly and reassembly of independent components. This flexibility caters to the diverse requirements of container environments.

**- Security:**

With a strong emphasis on security, Containerd implements features like user namespaces and seccomp profiles. These measures enhance container isolation, ensuring a robust security posture.

**- Compatibility:**

> Aligned with the Open Container Initiative (OCI) specifications, Containerd ensures compatibility with other runtimes and tools adhering to the OCI standard. This compatibility facilitates easy transitions between container runtimes supporting OCI.

Reviewer 2: ```suggestion
Aligned with the Open Container Initiative (OCI) specifications, containerd ensures compatibility with other runtimes and tools adhering to the OCI standard. This compatibility facilitates easy transitions between container runtimes supporting OCI.
```

Ensure the correct name `containerd` is used everywhere that it's not starting a sentence. See https://www.docker.com/blog/what-is-containerd-runtime/ for an example

**- CLI and APIs:**

Containerd provides well-defined APIs for programmatic interaction with container runtimes. Additionally, its Command-Line Interface (CLI) allows manual management of containers and images.

**- Production Ready:**

Widely adopted in multiple container orchestration platforms and cloud-native environments, Containerd has proven itself as a production-ready solution. Its reliability is evidenced by its integration into various deployments of containerized applications.

**- Community and Governance:**

As an open-source project under the Cloud Native Computing Foundation (CNCF), Containerd benefits from a diverse community of contributors. This collaborative approach ensures transparent decision-making, promoting inclusiveness and continuous improvement.

### Actors

**- Containerd Core:**

> Role: Serves as the core orchestration engine, managing the complete lifecycle of containers.

Reviewer 2: ```suggestion
Role: the core orchestration engine, managing the complete lifecycle of containers.
```

Functionality: Coordinates tasks such as image storage, transfer, execution, and supervision. Ensures a consistent and efficient containerized application experience.
Isolation: Adopts a modular design, separating concerns to prevent unauthorized access and actions. Implements access controls to reinforce security.

**- Container Runtimes:**

Role: Executes containers based on specifications provided by containerd, interacting directly with the underlying operating system.
Functionality: Translates container configurations into running instances, ensuring compatibility and adherence to standards.
Isolation: Operates within well-defined boundaries, utilizing namespaces and cgroups for robust process and resource isolation.

**- Image Registries:**

> Role: Acts as repositories for container images, collaborating with containerd in tasks such as image pulling, pushing, and managing metadata.

Reviewer 2: ```suggestion
Role: Repositories for container images, providing storage and retrieval for containerd in tasks such as image pulling, pushing, and metadata management.
```

**- Scope Clarity:**

The project scope is clearly defined, and any changes require a 100% vote from all maintainers. The whitelist approach ensures that anything not mentioned in scope is considered out of scope.

### Non-goals

**- Component Tight Coupling:**

Components should not have tight dependencies, promoting independence.

**- High-Level Abstractions in API:**

Avoid building high-level abstractions in the API, focus on exposing primitives.

**- Acceptance of Additional Implementations:**

Additional implementations for core components should not be accepted into the core repository and should be developed separately.

**- Build as a First-Class API:**

Building images is considered a higher-level feature and is out of scope.

**- Volume Management:**

Volume management for external data is out of scope. The API supports mounts, binds, etc., allowing different volume systems to be built on top.

**- Logging Persistence:**

Logging persistence is considered out of scope. Clients can handle and persist container STDIO as needed.

## Self-assessment use

> This self-assessment was authored by Swati Baleri, Vivek Radhakrishnan, Swati Baleri, Sunny Li, and Nathan Smith with a format established by the Containerd maintainers. The purpose of this document is to perform an internal analysis of the project's security.  It is not intended to provide a security audit of Containerd, or function as an independent assessment or attestation of Containerd's security health.

Reviewer 2: ```suggestion
This self-assessment was authored by Swati Baleri, Vivek Radhakrishnan, Sunny Li, and Nathan Smith with a format established by the Containerd maintainers. The purpose of this document is to perform an internal analysis of the project's security.  It is not intended to provide a security audit of Containerd, or function as an independent assessment or attestation of Containerd's security health.
```

# Self-assessment
The Self-assessment is the initial document for projects to begin thinking about the
security of the project, determining gaps in their security, and preparing any security
documentation for their users. This document is ideal for projects currently in the
CNCF **sandbox** as well as projects that are looking to receive a joint assessment and
currently in CNCF **incubation**.

For a detailed guide with step-by-step discussion and examples, check out the free 
Express Learning course provided by Linux Foundation Training & Certification: 
[Security Assessments for Open Source Projects](https://training.linuxfoundation.org/express-learning/security-self-assessments-for-open-source-projects-lfel1005/).

> # Self-assessment outline

Reviewer: Please go back to ensure that no template text like this is included here or in the rest of the document

**- Image Registries:**

Role: Acts as repositories for container images, collaborating with containerd in tasks such as image pulling, pushing, and managing metadata.
Significance: Critical components for image distribution, storage, and retrieval, forming a pivotal part of the containerized ecosystem.

**- System Administrators:**

Role: Configures, monitors, and maintains containerd in the broader system context, overseeing its integration into the overall infrastructure.
Responsibilities: Involves setup, continuous monitoring, optimization, and troubleshooting of containerd to ensure seamless operation.

**- Developers/Contributors:**

Role: Actively contributes to the containerd project through codebase enhancements, bug fixes, and feature development.
Responsibilities: Shapes the evolution of containerd, addressing issues, introducing improvements, and ensuring the project's ongoing robustness.

**- End Users:**
Role: Leverage containerd for deploying, managing, and orchestrating containerized applications.
Interaction: Engage with containerd through various interfaces and tools, contributing to the widespread adoption and integration of containerized solutions.

> ### Actions

Reviewer 3: This section talks about the various functions/benefits of this project however it is meant to showcase the individual parts of your system that interact to provide the desired functionality.  
Ref: https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md#actors

Please update this section to include the right actors.

**- Scope Clarity:**

The project scope is clearly defined, and any changes require a 100% vote from all maintainers. The whitelist approach ensures that anything not mentioned in scope is considered out of scope.

### Non-goals

**- Component Tight Coupling:**

Components should not have tight dependencies, promoting independence.

**- High-Level Abstractions in API:**

Avoid building high-level abstractions in the API, focus on exposing primitives.

**- Acceptance of Additional Implementations:**

Additional implementations for core components should not be accepted into the core repository and should be developed separately.

**- Build as a First-Class API:**

Building images is considered a higher-level feature and is out of scope.

**- Volume Management:**

Volume management for external data is out of scope. The API supports mounts, binds, etc., allowing different volume systems to be built on top.

**- Logging Persistence:**

Logging persistence is considered out of scope. Clients can handle and persist container STDIO as needed.

## Self-assessment use

> This self-assessment is created by the Containerd team to perform an internal analysis of the project's security.  It is not intended to provide a security audit of Containerd, or function as an independent assessment or attestation of Containerd's security health.

Reviewer 3: Can you please specify at the top which member of the maintainer team acted as an author on this? (Or otherwise rephrase)

  
   The project achieved the following
     - Basics: 13/13 passed
     - Change Control: 9/9 passed
     - Reporting: 8/8 passed
     - Quality: 13/13 passed
     - Security: 16/16 passed
     - Analysis: 8/8 passed

* Case Studies:
  
   Demonstrates how Red Hat OpenShift, integrated with containerd, streamlines containerization adoption and simplifies Kubernetes management.
  
   https://swapnasagarpradhan.medium.com/install-a-kubernetes-cluster-on-rhel8-with-conatinerd-b48b9257877a

  Explores how containerd simplifies container management on Google Kubernetes Engine (GKE), Google Cloud's fully managed Kubernetes service.

  https://cloud.google.com/kubernetes-engine

  Delves into the integration of containerd with Amazon Elastic Container Service (ECS), Amazon Web Services' container orchestration service

  https://aws.amazon.com/blogs/containers/tag/containerd/

  Explores how containerd enables organizations to effectively manage containers on Azure Kubernetes Service (AKS), Microsoft Azure's managed Kubernetes service

  https://azure.microsoft.com/en-us/updates/generally-available-containerd-support-for-windows-in-aks/
  

> * Related Projects / Vendors:

Reviewer 3: Please elaborate how these project/vendors relate to this project, and what the key differences between them are.
